WIP: distribution interface improvements

stevvooe opened the pull request:
This commit sketches out a series of proposed interface improvements for
defining interactions between distribution components. Please see doc.go for a
general overview.

Signed-off-by: Stephen J Day <stephen.day@docker.com>

Connects to #159 
Closes #264

stevvooe commented:
cc @dmcgowan @icecrime @jlhawn @crosbymichael @jfrazelle @dmp42

jlhawn commented on registry.go:
```diff
@@ -1,13 +1,6 @@
-	"github.com/docker/distribution/manifest"
-	"golang.org/x/net/context"
-)
+import "golang.org/x/net/context"
```
looks like this isn't needed.

jlhawn replied:
I don't think the interface should rely on an external package like this. I see that it's used in the `Registry` interface below to get a repository, but wouldn't it make more sense for an implementation of `Registry` to decide if it's going to use `x/net/context`?

stevvooe replied:
@jlhawn The idea here is that we subscribe to `x/net/context` or we don't. Really, each method should start with a context argument. Perhaps, we could refactor this interface to be transaction oriented to capture that, making context an implementation detail of the transaction.

stevvooe replied:
@jlhawn Please see docker/distribution#156 for a discussion of this issue. No one responded.

dmcgowan commented on blobs.go:
```diff
@@ -0,0 +1,64 @@
+	// Exists returns true if the blob exists.
+	Exists(dgst digest.Digest) (bool, error)
+
+	// Get the layer identifed by TarSum.
```
by digest or hash identifier or content address, anything but TarSum really :smile_cat:

stevvooe replied:
:hankey:

dmcgowan commented on manifests.go:
```diff
@@ -0,0 +1,43 @@
+
+	// Dependencies returns a list of object on which this manifest depends.
+	// Typically, these are layers.
+	Dependencies() []Descriptor
```
Is the application runtime information intended to be retrieved as a dependency or deserialized from the blob?

stevvooe replied:
It's part of the blob. It's up to the application to abstract it but it may be deserialized as part of the ManifestService.

dmcgowan commented on manifests.go:
```diff
@@ -0,0 +1,43 @@
+	// AddDependency includes the dependency in the manifest. If the add
+	// fails, such as when adding an unsupported dependency, an error may be
+	// returned.
+	AddDependency(dependency Descriptor) error
```
Order should be defined in the comment on this function. The list of dependencies from Manifest must be returned in the same order in which Add was called.

stevvooe replied:
Proposed a possibility: https://github.com/docker/distribution/pull/62/files#r25213693.

dmcgowan commented on manifests.go:
```diff
@@ -0,0 +1,43 @@
+	AddDependency(dependency Descriptor) error
+}
+
+// ManifestService describes operations on image manifests.
```
If I understand the purpose of this interface correctly, it is just to provide a nice wrapper around the BlobService when the media type is manifest. What other functionality would these calls add?  For example besides doing a blob write of the serialized manifest, what else does Put accomplish?

stevvooe replied:
It encompasses operations that semantically involve a manifest. Put on `ManifestService

dmcgowan replied:
Instead of treating manifest as a one-off special media-type, does it make sense to have commit hooks on blobs that can be routed per media-type?

stevvooe replied:
It is not a one-off, special media type. There is semantic meaning to putting a manifest versus just putting a blob. The presence of dependencies is verified.

Please propose a different interface that keeps things simple and provides these semantics if you don't think this will work.

dmcgowan commented on signatures.go:
```diff
@@ -0,0 +1,29 @@
+	Put(dgst digest.Digest, signatures ...Signature) error
+
+	// Delete removes one or more signatures from the targeted digest.
+	Delete(dgst digest.Digest, signatures ...Signature) error
```
Is the intent that these signatures will be stored by their own digest? Trying to understand why passing in a full Blob interface would be necessary here.

stevvooe replied:
They are "associated" with the digest. While signatures do have their own digest, this interface deals in their targets.

dmcgowan replied:
The delete function is intended to be given the full signature blob then remove each individual signature by the digest of the signature? The signature objects should be able to use their descriptors just like any other blob.  I assume the idea behind making it a blob is to easily allow different signature media type (PGP, JWS). If that is the case, why not just pass in signatures as digests.

I think this is still clear that dgst is the target.
~~~go
Delete(dgst digest.Digest, signatures ...digest.Digest) error
~~~

stevvooe replied:
@dmcgowan This doesn't address your concerns but perhaps adjusting the naming will make it more appropriate:

```
// SignatureService provides the ability to get and set signatures for a
// particular object. One or more signatures can be associated with a blob.
type SignatureService interface {
	// Get returns the signatures associated with the provided digest.
	Get(dgst digest.Digest) ([]Signature, error)

	// Assign  one or more signatures to the specified digest.
	Assign(dgst digest.Digest, signatures ...Signature) error

	// Remove one or more signatures from the specified digest.
	Remove(dgst digest.Digest, signatures ...Signature) error
}
```

The issue with using digests is that `SignatureService` would be tied to an underlying `BlobService`. If this speaks in external blob references, ie digests, and the actual data, it makes it possible to have external signature storage.

We probably need to discuss this further.

dmcgowan commented on tags.go:
```diff
@@ -0,0 +1,42 @@
+	All() (Tag, error)
+
+	// Get retrieves the digest for the provided tag.
+	Get(name string) (Tag, error)
```
These tags should be able to be fetched by digest as well.  If tags are being used as the method of trust by an external trust system, it will never fetch content by name but only content address.  Since signatures are on the tags, an external system will consider the tag to be verified and use its content-address to start the trusted lookup.

stevvooe replied:
The tag service resolves from a name to a digest. Going the other way, one would digest a tag object. If one wants access to tag content as a blob, using the blob service is the right approach. I am not sure of a use case in which one would fetch a tag by its digest.

dmcgowan replied:
Mostly referring to this conversation https://github.com/docker/distribution/pull/179/files#r24955886
Just trying to leave that as an option, it may be desirable that the external trust system completely disregards the tags interface on the registry in place of its own system.

stevvooe replied:
If the user requires a tag fetched by id, they should go through the blob interface. Tags manage name to digest mappings and that is the extent of the abstraction here.

stevvooe commented on signatures.go:
```diff
@@ -0,0 +1,29 @@
+
+// SignatureService provides the ability to get and set signatures for a
+// particular object. One or more signatures can be associated with a blob.
+type SignatureService interface {
```
cc @diogomonica @NathanMcCauley Please take a peak at this interface.

stevvooe commented on blobs.go:
```diff
@@ -0,0 +1,64 @@
+
+	// Descriptor identifies the layer.
+	Descriptor() Descriptor
+}
```
@diogomonica @NathanMcCauley Perhaps, we have signatures hanging off the blob itself (or maybe the tag).

icecrime commented on blobs.go:
```diff
@@ -0,0 +1,64 @@
+	// Exists returns true if the blob exists.
+	Exists(dgst digest.Digest) (bool, error)
+
+	// Get the layer identifed by digest dgst.
```
If I understand correctly, the name "layer" in this comment is badly chosen (can include manifests too).

jlhawn replied:
yes, the interface used to be called "LayerService", so I think the comment just wasn't updated.

icecrime commented on descriptor.go:
```diff
@@ -0,0 +1,30 @@
+	// Repository optionally specifies a named repository from which this
+	// content should always be accessed. Discovery should be performed to
+	// resolve the location.
+	Repository string `json:"repository,omitempty"`
```
Mmm... I'm trying to understand how this relates to the namespace document that was shared earlier. Isn't the discovery mechanism specific to a "top-level name" (i.e., an image)?

stevvooe replied:
The value of this field would be considered a "namespace", which points to a particular repository. The location of the repository and the registry required to fetch it would be resolved through the discovery process in the namespace proposal.

I'm worried that the term "namespace" is just too vague.

icecrime commented on doc.go:
```diff
@@ -1,7 +1,48 @@
+// Manifest
+//
+// A manifest is a blob that depends on other blobs. Effectively, the
+// repository describes trees of interelated objects. The manifest is the
```
interrelated

icecrime commented on tags.go:
```diff
@@ -0,0 +1,42 @@
+	// name will be "stevvooe/test:latest".
+	Name() string
+
+	// Description returns a human-readable description of the tag.
```
Is that so different from the `Name() string`?

stevvooe replied:
name: "stevvooe/test:latest"
description: "stevvooe/test:latest is an example used to demonstrate the return value of the `Name` method on the `Tag` interface. It is not a real image name. This description demonstrates the value of a human readable description that may help to identify the semantic importance of the tag. It is debatable whether or not this field is necessary."

icecrime commented on manifests.go:
```diff
@@ -0,0 +1,49 @@
+	// Dependencies returns a list of object on which this manifest depends.
+	// The dependencies are strictly ordered from base to head. Typically,
+	// these are layers but their interpretation is application specific.
+	Dependencies() []Descriptor
```
So if I want to "walk" the manifest tree, I must:

1. Iterate on its `Dependencies()`
2. Call `BlobService.Get(dependency.Digest)` to fetch the dependency blob
3. Switch on `dependencyBlob.Descriptor().MediaType` to determine what it is? Or `manifest ok := dependencyBlob.(Manifest)`?

To rephrase: I don't clearly see if what I fetch will be base blobs which `MediaType` I must examine, or if the library will do the job of recognizing a few mediaTypes (such as the one(s?) that identify a manifest) to instantiate `Blob`-descendent types (such as a `Manifest`).

stevvooe replied:
@icecrime After some thought, I think the answer is both. A raw blob service shouldn't have to know about its contents. This should be something we can support. However, if a use case requires traversal, one should be able to wrap the blob service in a component that detect's media types and returns interesting objects, such as a manifest, that could be switched upon.

jlhawn commented on blobs.go:
```diff
@@ -0,0 +1,64 @@
+	// against the descriptor, which may result in an error. The only required
+	// field for the descriptor is the digest and mediatype. If the size is
+	// present, it will be checked as well.
+	Commit(desc Descriptor) (Blob, error)
```
It would be convenient if all of Descriptor fields were optional and if omitted or empty the BlobWriter should use "application/octet-stream" as the default media type and just set the digest and size that it computed.

jlhawn commented on blobs.go:
```diff
@@ -0,0 +1,64 @@
+	io.Closer
+
+	// UUID returns the identifier for this writer.
+	UUID() string
```
Do we want to enforce that it's a UUID? or would a sufficiently random/unique string be okay?

stevvooe replied:
It's not enforced, in that the type is a string. We can call this `ID`, if that makes it more generalized.

stevvooe commented:
@jlhawn I've removed the repository field from descriptor, per our discussion in https://github.com/docker/distribution/pull/62/files#r25635220. Please update the manifest proposal accordingly.

stevvooe commented on tags.go:
```diff
@@ -0,0 +1,47 @@
+}
+
+// TagService provides access to information about tagged objects.
+type TagService interface {
```
Let's simplify the tag service to be a mapping between tag name to a `Descriptor`.

stevvooe commented on tags.go:
```diff
@@ -0,0 +1,47 @@
+
+// Tag is a named reference to a repository object. Target generally refers to
+// a manifest.
+type Tag interface {
```
This should go away.

stevvooe commented on tags.go:
```diff
@@ -0,0 +1,47 @@
+// TagService provides access to information about tagged objects.
+type TagService interface {
+	All() (Tag, error)
+
```
We need to add an `ErrUntrustedDescriptor` that exposes the descriptor but as an error. The caller can still lookup the target but they have to explicitly do so in an error handler. The initial implementations will probably use this path exclusively.

stevvooe commented:
- Added context to all service methods where appropriate
- Simplified tag service to operate with just names and descriptors
- Added `ErrTagUntrusted` to identify untrusted tags returned from `TagService`.

cc @dmcgowan @endophage @jlhawn @NathanMcCauley @diogomonica

dmcgowan commented:
LGTM!

dmcgowan commented:
If the tag service is handled through TUF, then the Put and Delete actions may require an explicit commit

dmp42 commented:
cc @docker/distribution-maintainers (<- just a test - is this working?)

stevvooe commented:
@dmcgowan Could that be managed by a transactional service? For example, does it matter if one gets `TagService` from a transactional service provider or a non-transactional provider? Imagine something like this:

```
tags, tx, err := StartTagTransaction()
// ... do things
tags.Get(...)
tags.Put(...)
tx.Commit()
```

Alternatively, one could wrap the entire set of repository operations:

```
repo, tx, err := StartRepositoryTransaction()
blobs, err := repo.Blobs()
//.. use blobs ...
tags, err := repo.Tags()
tx.Commit()
```

This approach separates transactions from the services. Unfortunately, it does avoid standardizing it.

Thoughts?

dmcgowan commented:
@stevvooe I think that is a fair way to look at it, since this is the top level interface it will likely get wrapped by the final implementations in such functionality.  Let's re-address in a few weeks when we have a firmer picture of the Trust architecture, any change I would image is small or additive.

stevvooe commented:
@dmcgowan Added the `Namespace` object.

dmcgowan commented on registry.go:
```diff
@@ -1,20 +1,36 @@
+// Scope defines the set of items that match a namespace.
+type Scope interface {
+	// Contains returns true if the name belongs to the namespace.
+	Contains(name string) (bool, error)
```
What's the case you would expect this to throw an error if its just doing a string comparison?

stevvooe replied:
We can probably remove this. It really depends on how we define scope. If it is just a syntax-based set (ie, glob or equiv), then we don't need the error here.

What if the implementation has to ask a registry?

dmcgowan replied:
If the Scope object needs to make a request elsewhere, its implementation is more complex than Scope should be.  This should really be a simple string function, not a membership check.

stevvooe replied:
Agreed.

dmcgowan commented:
+1 to Namespace change, when do you foresee this PR getting ready to merge?

stevvooe commented:
@dmcgowan When 1.6 settles, this is going to be my primary goal. Realistically, we'll have to phase these changes in. For instance, changing `Registry` to `Namespace` could be a relatively small PR.

dmcgowan commented on registry.go:
```diff
@@ -1,20 +1,36 @@
+	// Index returns an IndexService that provides search across the
+	// namespace, supporting content discovery and collective information
+	// about available images.
+	// Index(ctx context.Context) (IndexService, error)
```
We should take this into consideration now since we are defining the registry service as an implementation of Namespace.  If namespace covers more than repositories (which I think we agree it does), then what does it mean for Registry to be a partial implementation?  Should a registry throw an error when an Index is retrieved, or should Namespace be a composed type.

stevvooe replied:
@dmcgowan Right now, the search/index proposal is too new. I think we can make this decision in the future. A compositional type with interfaces might make sense, but let's get a few in place before going too far.

ncdc commented:
@stevvooe do you anticipate this going in for 2.0?

stevvooe commented:
@ncdc No. This will happen after 2.0.

I am going to stage the changes in, starting with the blob store and working my way up.

stevvooe commented on blobs.go:
```diff
@@ -0,0 +1,71 @@
+	io.ReaderAt
+	io.Closer
+
+	Reader() (ReadSeekCloser, error)
```
@dmcgowan @tibor What do you think about removing this method completely and just supporting io.ReaderAt on blob. It allows us to abstract blobs as a series of distributed chunks and defers the `io.Reader` implementation to the caller (ie `NewReadSeeker(length int64, blob io.ReaderAt) (io.ReadCloser, error)`).

dmcgowan replied:
Why not `NewReadSeeker(blob io.ReaderAt) (io.ReadSeekCloser, error)` then every read becomes a ReadAt and the underlying ReadSeeker will need to keep track of the read state.

dmcgowan replied:
I generally like the idea as long as we are not losing any optimizations from Seek + ReadAll

RichardScothern commented on descriptor.go:
```diff
@@ -0,0 +1,29 @@
+	// against against this digest.
+	Digest digest.Digest `json:"digest,omitempty"`
+
+	// NOTE(stevvooe): Before adding a field here, please ensure that all
```
If we are going with the notion that content is immutable I think this struct have un-exported fields.  Being able to mutate the fields seems a bit odd, and unnecessary.

stevvooe replied:
@RichardScothern Perhaps, this should be an interface?

RichardScothern replied:
I'm not sure that's necessary; a blob descriptor as a struct makes sense.  Like the blob it describes, I think it makes sense to be immutable.

stevvooe replied:
@RichardScothern An interface with only getters is the best way to _enforce_ immutability. For immutable data records, without enforcement, these struct definitions are the best. Just ensure that they are always passed by value, rather than as a pointer.

vbatts commented:
/cc @vbatts

vbatts commented on blobs.go:
```diff
@@ -0,0 +1,71 @@
+type BlobWriter interface {
+	io.WriteSeeker
+	io.WriterAt
+	io.ReaderFrom // TODO(stevvooe): This should be optional.
```
can't you just not put it here, and level up with a 
```golang
type BlobWriterFrom interface {
    BlobWriter
    io.ReaderFrom
}
```

dmcgowan replied:
> // TODO(stevvooe): This should be optional.

`io.ReaderFrom` is probably the most important optimization from the client perspective. Without it either making a request per write is necessary or having an open request in which Write sends to the request body.

vbatts replied:
Understandable. Then let's drop the comment? :-)
On May 6, 2015 9:28 AM, "Derek McGowan" <notifications@github.com> wrote:

> In blobs.go
> <https://github.com/docker/distribution/pull/193#discussion_r29775602>:
>
> > +	io.Closer
> > +
> > +	Reader() (ReadSeekCloser, error)
> > +
> > +	// Handler returns an HTTP handler which serves the blob content, either
> > +	// by redirection or directly serving the content itself.
> > +	Handler(r *http.Request) (http.Handler, error)
> > +}
> > +
> > +// BlobWriter provides a handle for working with a blob store. Instances
> > +// should be obtained from BlobService.Writer and BlobService.Resume. If
> > +// supported by the store, a writer can be recovered with the id.
> > +type BlobWriter interface {
> > +	io.WriteSeeker
> > +	io.WriterAt
> > +	io.ReaderFrom // TODO(stevvooe): This should be optional.
>
>  // TODO(stevvooe): This should be optional.
>
> io.ReaderFrom is probably the most important optimization from the client
> perspective. Without it either making a request per write is necessary or
> having an open request in which Write sends to the request body.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/distribution/pull/193/files#r29775602>.
>

stevvooe replied:
@vbatts Agreed.

stevvooe commented:
I believe this is being picked up by @RichardScothern.
